persistence-api/src/main/java/io/stargate/db/Persistence.java

```diff
@@ -15,96 +15,162 @@
+   * Registers a listener whose methods will be called when either the schema or the topology of the
+   * persistence changes.
    */
-  Schema schema();
-
   void registerEventListener(EventListener listener);
 
   boolean isRpcReady(InetAddressAndPort endpoint);
 
   InetAddressAndPort getNativeAddress(InetAddressAndPort endpoint);
 
-  QueryState<Q> newQueryState(ClientState<C> clientState);
-
-  ClientState<C> newClientState(SocketAddress remoteAddress, InetSocketAddress publicAddress);
-
-  ClientState newClientState(String name);
-
-  AuthenticatedUser<?> newAuthenticatedUser(String name);
-
   Authenticator getAuthenticator();
 
-  DataStore newDataStore(QueryState<Q> state, QueryOptions queryOptions);
+  /**
+   * Creates a new connection for an "external" client identified by the provided info.
+   *
+   * @param clientInfo info identifying the client for which the connection is created.
+   * @return the newly created persistence connection.
+   */
+  Connection newConnection(ClientInfo clientInfo);
+
+  /**
+   * Creates a new "internal" connection, that is one with no client information.
+   *
+   * @return the newly created persistence connection.
+   */
+  Connection newConnection();
 
   /**
    * The object that should be used to act as an 'unset' value for this persistence (for use in
-   * {@link QueryOptions#getValues()} or in the {@code values} argument of {@link #batch}).
+   * {@link Statement#values()}).
    *
    * <p>Please note that persistence implementations are allowed to use <b>reference equality</b> to
    * detect this value, so the object returned by this method should be used "as-is" and should
    * <b>not</b> be copied (through {@link ByteBuffer#duplicate()} or any other method).
    */
   ByteBuffer unsetValue();
 
-  CompletableFuture<? extends Result> query(
-      String cql,
-      QueryState state,
-      QueryOptions options,
-      Map<String, ByteBuffer> customPayload,
-      boolean isTracingRequested,
-      long queryStartNanoTime);
-
-  CompletableFuture<? extends Result> execute(
-      MD5Digest id,
-      QueryState state,
-      QueryOptions options,
-      Map<String, ByteBuffer> customPayload,
-      boolean isTracingRequested,
-      long queryStartNanoTime);
-
-  CompletableFuture<? extends Result> prepare(
-      String cql,
-      QueryState state,
-      Map<String, ByteBuffer> customPayload,
-      boolean isTracingRequested);
-
-  CompletableFuture<? extends Result> batch(
-      BatchType type,
-      List<Object> queryOrIds,
-      List<List<ByteBuffer>> values,
-      QueryState state,
-      QueryOptions options,
-      Map<String, ByteBuffer> customPayload,
-      boolean isTracingRequested,
-      long queryStartNanoTime);
-
   boolean isInSchemaAgreement();
 
-  void captureClientWarnings();
-
-  List<String> getClientWarnings();
+  /** Wait for schema to agree across the cluster */
+  default void waitForSchemaAgreement() {
+    for (int count = 0; count < 100; count++) {
```

Comment:
Could we change the timeout into a parameter? (perhaps add another method that take a timeout parameter?)

Reply:
... and have a system property to control the default timeout perhaps?

Reply:
That method is, contrary to what it looks like, not new to this patch; it's just moved as-is from `DataStore`. And I didn't wanted to mess up with schema agreement in that patch so I left it as is. Which isn't to say I disagree in principle, and in fact I need to dig why we have both this method and the different (but essentially doing the same thing) `SchemaAgreement` class in the `cql` module.

So if that's ok, I'll create a quick follow-up issue to clean up schema agreement code.

Reply:
sure. thx!